pkg/agent/route/route_windows.go

```diff
@@ -280,28 +280,43 @@ func (c *Client) addServiceCIDRRoute(serviceCIDR *net.IPNet) error {
 	c.serviceRoutes.Store(serviceIPv4CIDRKey, route)
 
 	// Collect stale routes.
-	var staleRoutes []*util.Route
-	// If current destination CIDR is not nil, the route with current destination CIDR should be uninstalled since
-	// a new route with a newly calculated destination CIDR has been installed.
+	staleRoutes := make(map[string]*util.Route)
 	if serviceCIDRRouteExists {
-		staleRoutes = append(staleRoutes, oldServiceCIDRRoute.(*util.Route))
+		// If current destination CIDR is not nil, the route with current destination CIDR should be uninstalled since
+		// a new route with a newly calculated destination CIDR has been installed.
+		rt := oldServiceCIDRRoute.(*util.Route)
+		staleRoutes[rt.String()] = rt
+	} else {
+		// If current destination CIDR is nil, which means that Antrea Agent has just started, then all existing routes
+		// whose destination CIDR contains the first ClusterIP should be uninstalled, except the newly installed route.
+		// Note that, there may be multiple stale routes prior to this commit. When upgrading, all stale routes will be
+		// collected. After this commit, there will be only one stale route after Antrea Agent started.
+		routes, err := c.listIPRoutesOnGW()
+		if err != nil {
+			return fmt.Errorf("error listing ip routes: %w", err)
+		}
+		for _, rt := range routes {
+			if rt.GatewayAddress.Equal(gw) && !rt.DestinationSubnet.IP.Equal(serviceCIDR.IP) && rt.DestinationSubnet.Contains(serviceCIDR.IP) {
+				staleRoutes[rt.String()] = rt
+			}
+		}
 	}
 	routes, err := c.listIPRoutesOnGW()
 	if err != nil {
 		return fmt.Errorf("error listing ip routes: %w", err)
 	}
 	// Collect stale per-IP routes for ClusterIPs before this patch.
+	// TODO(lhongliang): remove this in v1.13.0 or later.
 	for _, rt := range routes {
-		ones, _ := rt.DestinationSubnet.Mask.Size()
-		if ones == net.IPv4len*8 && serviceCIDR.Contains(rt.DestinationSubnet.IP) {
-			staleRoutes = append(staleRoutes, rt)
+		if iputil.IsSubCIDR(serviceCIDR, rt.DestinationSubnet) {
+			staleRoutes[rt.String()] = rt
```

Comment:
why it lists routes twice? couldn't get the undesired routes in the above else branch by a unified condition?
```
// Not the routes we are interested in.
if !rt.GatewayAddress.Equal(gw) {
        continue
}
// It's the latest route we just installed.
if net.IPNetEqual(rt.DestinationSubnet, serviceCIDR) {
        continue
}
// The route covers the desired route. It was installed when the calculated ServiceCIDR is larger than the current one, which could happen after some Services are deleted.
if rt.DestinationSubnet.Contains(serviceCIDR.IP) {
        staleRoutes = append(staleRoutes, rt)
}
// The desired route covers the route. It was either installed when the calculated ServiceCIDR is smaller than the current one, or a per-IP route generated before v1.12.0.
if serviceCIDR.Contains(rt.DestinationSubnet.IP) {
        staleRoutes = append(staleRoutes, rt)
}
```
It doesn't need to change staleRoutes to map.

Reply:
It lists routes twice only when installing the first Service.  For the first listing of routes, it is used to consistent with Linux code. Besides, the logic is used to find the existing routes whose destinations contains the first ClusterIP (serviceCIDR.IP), and it is only called when syncing the first ClusterIP.

For the second listing of routes, it is used to collect the per-IP route generated before v1.12.0. The reason to change staleRoutes to map is that:
- Assuming the first ClusterIP is 10.96.0.1, and the second one is 10.96.0.2
- After syncing the second ClusterIP, the route for 10.96.0.1/32 will be added to `staleRoutes` according to L284-L288
- The route 10.96.0.1 will be also listed in `routes` in L304
- The route 10.96.0.1 will be added to `staleRoutes` according to L311-L314 again.

Reply:
1. They have different cases to handle and their code are different anyway, I don't feel it's important to use partially consistent code but sacrifice extra system calls.
2. It doesn't make sense to relist the routes everytime the ServiceCIDR is updated, I guess the problem is the implementation of the discover generating intermediate values. I think there are more problems in it, https://github.com/antrea-io/antrea/pull/5017 should fix it.
3. Even the Linux implementation seems wrong as it can only clean up larger CIDR but not the smaller ones.